core/consensus/babe/primitives/src/lib.rs

```diff
@@ -87,57 +92,55 @@ pub struct Epoch {
 	pub authorities: Vec<(AuthorityId, BabeAuthorityWeight)>,
 	/// Randomness for this epoch
 	pub randomness: [u8; VRF_OUTPUT_LENGTH],
-	/// Whether secondary slot assignments should be used during the epoch.
-	pub secondary_slots: bool,
 }
 
 /// An consensus log item for BABE.
 #[derive(Decode, Encode, Clone, PartialEq, Eq)]
 pub enum ConsensusLog {
-	/// The epoch has changed. This provides information about the
-	/// epoch _after_ next: what slot number it will start at, who are the authorities (and their weights)
-	/// and the next epoch randomness. The information for the _next_ epoch should already
-	/// be available.
+	/// The epoch has changed. This provides information about the _next_
+	/// epoch - information about the _current_ epoch (i.e. the one we've just
+	/// entered) should already be available earlier in the chain.
 	#[codec(index = "1")]
-	NextEpochData(Epoch),
+	NextEpochData(NextEpochDescriptor),
 	/// Disable the authority with given index.
 	#[codec(index = "2")]
 	OnDisabled(AuthorityIndex),
 }
 
 /// Configuration data used by the BABE consensus engine.
-#[derive(Copy, Clone, Hash, PartialEq, Eq, Debug, Encode, Decode)]
+#[derive(Clone, PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(any(feature = "std", test), derive(Debug))]
 pub struct BabeConfiguration {
 	/// The slot duration in milliseconds for BABE. Currently, only
 	/// the value provided by this type at genesis will be used.
 	///
 	/// Dynamic slot duration may be supported in the future.
 	pub slot_duration: u64,
 
+	/// The duration of epochs in slots.
+	pub epoch_length: SlotNumber,
+
 	/// A constant value that is used in the threshold calculation formula.
-	/// Expressed as a fraction where the first member of the tuple is the
-	/// numerator and the second is the denominator. The fraction should
+	/// Expressed as a rational where the first member of the tuple is the
+	/// numerator and the second is the denominator. The rational should
 	/// represent a value between 0 and 1.
 	/// In the threshold formula calculation, `1 - c` represents the probability
 	/// of a slot being empty.
 	pub c: (u64, u64),
 
-	/// The minimum number of blocks that must be received before running the
-	/// median algorithm to compute the offset between the on-chain time and the
-	/// local time. Currently, only the value provided by this type at genesis
-	/// will be used, but this is subject to change.
-	///
-	/// Blocks less than `self.median_required_blocks` must be generated by an
-	/// *initial validator* ― that is, a node that was a validator at genesis.
-	pub median_required_blocks: u64,
+	/// The authorities for the genesis epoch.
+	pub genesis_authorities: Vec<(AuthorityId, BabeAuthorityWeight)>,
+
+	/// The randomness for this epoch.
+	pub randomness: [u8; VRF_OUTPUT_LENGTH],
+
+	/// Whether this chain should run with secondary slots, which are assigned
+	/// in round-robin manner.
+	pub secondary_slots: bool,
```

Comment:
Another option might be `pub secondary_slots: u16,` or `Option<..>` that expresses how many VRF slots should elapse between secondary slots, with `u16::MAX` or `None` denoting not to bother with secondary slots.